### PR TITLE
Added the ability to recreate the PS stream

### PR DESCRIPTION
**Pull request by derwind**
This is a pull request for #2503.

# Changes

Recreate the  PS stream if necessary during `saveAs`.

# Description

- To avoid affecting the existing behavior, I make sure that the additionally implemented codes will not be executed unless `parse` is executed and `saveAs` is called.
- Some implementations were not derived purely from the Type 1 standards, but were based on AFDKO, so there may be some bias.
- In addition, the Type 1 fonts available were limited for me, so the verifications by me are also limited.

### CODE REVIEW

File: Lib/fontTools/t1Lib/__init__.py [hunk 1]
```diff
@@ -367,3 +513,72 @@ def stringToLong(s):
...
+font_dictionary_keys.remove("FontMatrix")
+
+FontInfo_dictionary_keys = copy.copy(_type1_fontinfo_order)
+# extend because AFDKO tx may use following keys
+FontInfo_dictionary_keys.extend([
+	"FSType",
+	"Copyright",
+])
+
+Praivate_dictionary_keys = [
```
Comment (justvanrossum) [inline Lib/fontTools/t1Lib/__init__.py:532]: Typo in Private_dictionary_keys

File: Tests/t1Lib/t1Lib_test.py [hunk 2]
```diff
@@ -52,23 +55,40 @@ def test_read_pfa_write_pfb(self):
...
+		saved_font = self.write(font, 'OTHER', dohex=True, doparase=True)
+		self.assertTrue(same_dicts(font.font, saved_font))
+
 	def test_read_with_path(self):
 		import pathlib
 		font = t1Lib.T1Font(pathlib.Path(PFB))
 
 	@staticmethod
-	def write(font, outtype, dohex=False):
+	def write(font, outtype, dohex=False, doparase=False):
```
Comment (justvanrossum) [inline Tests/t1Lib/t1Lib_test.py:82]: `doparase` -> `doparse` ?

File: Lib/fontTools/t1Lib/__init__.py [hunk 3]
```diff
@@ -367,3 +513,72 @@ def stringToLong(s):
 	for i in range(4):
 		l += byteord(s[i]) << (i * 8)
 	return l
+
+
+# PS stream helpers
+
+font_dictionary_keys = copy.copy(_type1_pre_eexec_order)
```
Comment (justvanrossum) [inline Lib/fontTools/t1Lib/__init__.py:520]: no need to use the copy module here, you can just do `list(_type1_pre_eexec_order)`

File: Lib/fontTools/t1Lib/__init__.py [hunk 4]
```diff
@@ -367,3 +513,72 @@ def stringToLong(s):
...
+hintothers = \
+	"/OtherSubrs[{}{}{}{systemdict/internaldict known not{pop 3}{1183615869\n" + \
+	"systemdict/internaldict get exec dup/startlock known{/startlock get exec}{dup\n" + \
+	"/strtlck known{/strtlck get exec}{pop 3}ifelse}ifelse}ifelse}executeonly]def"
+# t1write.c:saveStdSubrs
+def bytechrjoin(iterable, joiner=b""):
+	return tobytes(joiner).join(bytechr(item) for item in iterable)
+std_subrs = [
+	# 3 0 callother pop pop setcurrentpoint return
+	bytechrjoin([0x8e, 0x8b, 0xc, 0x10, 0xc, 0x11, 0xc, 0x11, 0xc, 0x21, 0xb]),
```
Comment (justvanrossum) [inline Lib/fontTools/t1Lib/__init__.py:569]: Why not use byte literals, like you do below?

File: Lib/fontTools/t1Lib/__init__.py [hunk 5]
```diff
@@ -367,3 +513,72 @@ def stringToLong(s):
...
+	"ForceBold",
+	"LanguageGroup",
+	"password",
+	"lenIV",
+	"MinFeature",
+	"RndStemUp",
+]
+
+# t1write_hintothers.h
+hintothers = \
```
Comment (justvanrossum) [inline Lib/fontTools/t1Lib/__init__.py:560]: This may look nicer as a triple-quoted string.

**Review by justvanrossum — COMMENTED** [review]
First round of feedback.

File: Lib/fontTools/t1Lib/__init__.py (hunk 1 above)
Comment (derwind) [reply Lib/fontTools/t1Lib/__init__.py:532]: Thank you. I'll fix it.

File: Tests/t1Lib/t1Lib_test.py (hunk 2 above)
Comment (derwind) [reply Tests/t1Lib/t1Lib_test.py:82]: Yes... I'll fix them.

File: Lib/fontTools/t1Lib/__init__.py (hunk 4 above)
Comment (derwind) [reply Lib/fontTools/t1Lib/__init__.py:569]: OK. I copied them from [t1write.c](https://github.com/adobe-type-tools/afdko/blob/develop/c/shared/source/t1write/t1write.c#L590-L605) and used them, which is why they look like this.

File: Lib/fontTools/t1Lib/__init__.py (hunk 3 above)
Comment (derwind) [reply Lib/fontTools/t1Lib/__init__.py:520]: OK, I will use `list()`. I checked the [Python 3.7 documents](https://docs.python.org/3.7/library/stdtypes.html#typesseq-list) just to be sure, and it says that list() will make a copy if _iterable_ is already a list.

File: Lib/fontTools/t1Lib/__init__.py (hunk 5 above)
Comment (derwind) [reply Lib/fontTools/t1Lib/__init__.py:560]: I'll use a triple-quoted string.

**Comment by justvanrossum** [comment]
Looks ok to me. @anthrotype, what do you think? Given the small audience this will have, I'm tempted to just merge and not worry too much about further fine tuning the code.

**Comment by anthrotype** [comment]
Sounds good to me, thanks!

**Pull request merged**